Add column types to `ActiveRecord::Result` for SQLite3

ankane opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

This allows `ActiveRecord::Result#cast_values` to cast dates and times with SQLite like it does with PostgreSQL. This is useful for queries that use `ActiveRecord::Base.lease_connection.select_all`. Column types can be referenced by column name or position like with PostgreSQL.

Relevant PostgreSQL adapter code: https://github.com/rails/rails/blob/v8.0.1/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb#L177-L184

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

ankane committed:
Add column types to `ActiveRecord::Result` for SQLite3

byroot commented on activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:
```diff
@@ -100,7 +104,11 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
                   ActiveRecord::Result.empty
                 else
-                  ActiveRecord::Result.new(stmt.columns, stmt.to_a)
+                  types = {}
+                  stmt.columns.zip(stmt.types).each_with_index do |(c, t), i|
+                    types[c] = types[i] = type_map.lookup(t)
+                  end
+                  ActiveRecord::Result.new(stmt.columns, stmt.to_a, types.freeze)
```
I get that you copied what was done for PG, but this implementation isn't great performance wise. I probably never noticed it because I never profile with pg.

I'll experiment a few things.

byroot committed:
Lazily build `ActiveRecord::Result#column_types`

Given this attribute is very rarely accessed, delaying as many
computations as possible is preferable.

ankane commented:
Thanks @byroot!
